README.md
parameter
=======

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![npm download][download-image]][download-url]
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg?style=flat-square)](https://github.com/semantic-release/semantic-release)

[npm-image]: https://img.shields.io/npm/v/parameter.svg?style=flat-square
[npm-url]: https://npmjs.org/package/parameter
[travis-image]: https://img.shields.io/travis/node-modules/parameter.svg?style=flat-square
[travis-url]: https://travis-ci.org/node-modules/parameter
[codecov-image]: https://codecov.io/github/node-modules/parameter/coverage.svg?branch=master
[codecov-url]: https://codecov.io/github/node-modules/parameter?branch=master
[download-image]: https://img.shields.io/npm/dm/parameter.svg?style=flat-square
[download-url]: https://npmjs.org/package/parameter

A parameter verify tools.

## Install

```bash
$ npm install parameter --save
```

## Usage

### API

`Parameter` Class

- `constructor([options])` - new Class `Parameter` instance
  - `options.translate` - translate function
  - `options.validateRoot` - config whether to validate the passed in value must be a object, default to `false`.
  - `options.convert` - convert primitive params to specific type, default to `false`.
  - `optinos.widelyUndefined` - convert empty string(`''`), NaN, Null to undefined, this option can make `rule.required` more powerful, default to `false`.__This may change the original input params__.
- `validate(rule, value)` - validate the `value` conforms to `rule`. return an array of errors if break rule.
- `addRule(type, check)` - add custom rules.
   - `type` - rule type, required and must be string type.
   - `check` - check handler. can be a `function` or a `RegExp`.

__Note: when `options.convert` enabled, all built-in rules check for primitive input param and convert it to rule's default `convertType`(which defined below), you can also enable this feature for specific rule by `convertType` options in each rule definition.__

### Example

```js
var Parameter = require('parameter');

var parameter = new Parameter({
  translate: function() {
    var args = Array.prototype.slice.call(arguments);
    // Assume there have I18n.t method for convert language.
    return I18n.t.apply(I18n, args);
  },
  validateRoot: true, // restrict the being validate value must be a object
});

var data = {
  name: 'foo',
  age: 24,
  gender: 'male'
};

var rule = {
  name: 'string',
  age: 'int',
  gender: ['male', 'female', 'unknown']
};

var errors = parameter.validate(rule, data);
```

#### [complex example](example.js)

### Rule

#### common rule

- `required` - if `required` is set to false, this property can be null or undefined. default to `true`.
- `type` - The type of property, every type has it's own rule for the validate.
- `convertType` - Make parameter convert the input param to the specific type, support `int`, `number`, `string` and `boolean`, also support a function to customize your own convert method.
- `default` - The default value of property, once the property is allowed non-required and missed, parameter will use this as the default value. __This may change the original input params__.
- `widelyUndefined` - override `options.widelyUndefined`

__Note: you can combile require and type end with a notation `?` like: `int?` or `string?` to specific both type and non-required.__

#### int

If type is `int`, there has tow addition rules:

- `max` - The maximum of the value, `value` must <= `max`.
- `min` - The minimum of the value, `value` must >= `min`.

Default `convertType` is `int`.

__Note: default `convertType` will only work when `options.convert` set to true in parameter's constructor.__

#### integer

Alias to `int`.

#### number

If type is `number`, there has tow addition rules:

- `max` - The maximum of the value, `value` must <= `max`.
- `min` - The minimum of the value, `value` must >= `min`.

Default `convertType` is `number`.

#### date

The `date` type want to match `YYYY-MM-DD` type date string.

Default `convertType` is `string`.

#### dateTime

The `dateTime` type want to match `YYYY-MM-DD HH:mm:ss` type date string.

Default `convertType` is `string`.

#### datetime

Alias to `dateTime`.

#### id

The `id` type want to match `/^\d+$/` type date string.

Default `convertType` is `string`.

#### boolean

Match `boolean` type value.

Default `convertType` is `boolean`.

#### bool

Alias to `boolean`

#### string

If type is `string`, there has four addition rules:

- `allowEmpty`(alias to `empty`) - allow empty string, default to false. If `rule.required` set to false, `allowEmpty` will be set to `true` by default.
- `format` - A `RegExp` to check string's format.
- `max` - The maximum length of the string.
- `min` - The minimum length of the string.
- `trim` - Trim the string before check, default is `false`.

Default `convertType` is `string`.

#### email

The `email` type want to match [RFC 5322](http://tools.ietf.org/html/rfc5322#section-3.4) email address.

- `allowEmpty` - allow empty string, default is false.

Default `convertType` is `string`.

#### password

The `password` type want to match `/^$/` type string.

- `compare` - Compare field to check equal, default null, not check.
- `max` - The maximum length of the password.
- `min` - The minimum length of the password, default is 6.

Default `convertType` is `string`.

#### url

The `url` type want to match [web url](https://gist.github.com/dperini/729294).

Default `convertType` is `string`.

#### enum

If type is `enum`, it requires an addition rule:

- `values` - An array of data, `value` must be one on them. ___this rule is required.___

#### object

If type is `object`, there has one addition rule:

- `rule` - An object that validate the properties ot the object.

#### array

If type is `array`, there has four addition rule:

- `itemType` - The type of every item in this array.
- `rule` - An object that validate the items of the array. Only work with `itemType`.
- `max` - The maximun length of the array.
- `min` - The minimun lenght of the array.

#### abbr

- `'int'` => `{type: 'int', required: true}`
- `'int?'` => `{type: 'int', required: false }`
- `'integer'` => `{type: 'integer', required: true}`
- `'number'` => `{type: 'number', required: true}`
- `'date'` => `{type: 'date', required: true}`
- `'dateTime'` => `{type: 'dateTime', required: true}`
- `'id'` => `{type: 'id', required: true}`
- `'boolean'` => `{type: 'boolean', required: true}`
- `'bool'` => `{type: 'bool', required: true}`
- `'string'` => `{type: 'string', required: true, allowEmpty: false}`
- `'string?'` => `{type: 'string', required: false, allowEmpty: true}`
- `'email'` => `{type: 'email', required: true, allowEmpty: false, format: EMAIL_RE}`
- `'password'` => `{type: 'password', required: true, allowEmpty: false, format: PASSWORD_RE, min: 6}`
- `'object'` => `{type: 'object', required: true}`
- `'array'` => `{type: 'array', required: true}`
- `[1, 2]` => `{type: 'enum', values: [1, 2]}`
- `/\d+/` => `{type: 'string', required: true, allowEmpty: false, format: /\d+/}`

### `errors` examples

#### `code: missing_field`

```js
{
  code: 'missing_field',
  field: 'name',
  message: 'required'
}
```

#### `code: invalid`

```js
{
  code: 'invalid',
  field: 'age',
  message: 'should be an integer'
}
```

### Release process

We're using [semantic-release](https://github.com/semantic-release/semantic-release) to run npm publish
after every commit on master.

See [Default Commit Message Format](https://github.com/semantic-release/semantic-release#default-commit-message-format) for details.

## License

[MIT](LICENSE.txt)


History.md

3.6.0 / 2019-05-22
==================

**features**
  * [[`156fc36`](http://github.com/node-modules/parameter/commit/156fc3609710369b1a28f8618c4abf18111ac948)] - feat: support convertType function added by addRule (hdumok <<hdumok@163.com>>)

3.5.0 / 2018-12-11
==================

**features**
  * [[`ee3521c`](http://github.com/node-modules/parameter/commit/ee3521c74e73630b5b72014d7aeba7ed87eb865e)] - feat: support override options.widelyUndefined (#59) (Yiyu He <<dead_horse@qq.com>>)

3.4.0 / 2018-10-30
==================

**features**
  * [[`abcd739`](http://github.com/node-modules/parameter/commit/abcd739cd24133205e1f0c34e2d2a4adbf7be921)] - feat: support es5 with babel (#56) (Angela <<idu.angela@gmail.com>>)

3.3.1 / 2018-10-09
==================

**fixes**
  * [[`f5e33e3`](http://github.com/node-modules/parameter/commit/f5e33e379d0c70616faed6d010cf75b577818bb5)] - fix: allow default to be `null` or `false` (#55) (Cong Liu <<ghostoy@users.noreply.github.com>>)

3.3.0 / 2018-09-19
==================

**features**
  * [[`4d95c6a`](http://github.com/node-modules/parameter/commit/4d95c6a52416c8b8cd4f71b647ec4c05e0bab8ac)] - feat: support `trim: true` on string value (#54) (fengmk2 <<fengmk2@gmail.com>>)

3.2.0 / 2018-09-11
==================

**features**
  * [[`9d35b7f`](http://github.com/node-modules/parameter/commit/9d35b7fda2b21bf59d4ba036f7b75874d82ca9ad)] - feat: support options.widelyUndefined and rule.default (#53) (Yiyu He <<dead_horse@qq.com>>)

3.1.0 / 2018-08-30
==================

**features**
  * [[`487ae22`](http://github.com/node-modules/parameter/commit/487ae229df85b235761bfbb402826baaf61f29bb)] - feat: addRule support set default convertType (#51) (fengmk2 <<fengmk2@gmail.com>>)

3.0.0 / 2018-08-30
==================

**others**
  * [[`f1dda26`](http://github.com/node-modules/parameter/commit/f1dda26ac3697c63a7e344228b7aed7bb55b3931)] - refactor: make rules more practical (#50) (Yiyu He <<dead_horse@qq.com>>)

2.4.0 / 2018-08-03
==================

**features**
  * [[`e12fac4`](http://github.com/node-modules/parameter/commit/e12fac4d1681e51b02193defd5acb9e4864138dd)] - feat: add global addRule method (#46) (fengmk2 <<fengmk2@gmail.com>>)

2.3.0 / 2018-06-19
==================

**features**
  * [[`a5c1721`](http://github.com/node-modules/parameter/commit/a5c172179071aaaafb6fac604bc6bf4ebb0c8d98)] - feat: Add option.validateRoot to enable validate the root object (#41) (paranoidjk <<hust2012jiangkai@gmail.com>>)

2.2.3 / 2018-05-16
==================

2.2.2 / 2018-05-04
==================

**fixes**
  * [[`8317071`](http://github.com/node-modules/parameter/commit/83170710bbc88b1d7267cba6b9eae764a1e5fd36)] - fix: email regx to case insensitive (#36) (mansonchor.github.com <<mansonchor@126.com>>),

2.2.1 / 2017-11-23
==================

**fixes**
  * [[`3693861`](http://github.com/node-modules/parameter/commit/3693861897e2d207edf791f56db51820f76588b2)] - fix: check number  exclude NaN (#30) (Gao Peng <<ggjqzjgp103@qq.com>>)

**others**
  * [[`67d8cf6`](http://github.com/node-modules/parameter/commit/67d8cf6269abe4fdf3480b6c25cd08b67868ed11)] - chore(docs): add semantic-release process description (fengmk2 <<fengmk2@gmail.com>>)

2.2.0 / 2017-06-13
==================

  * feat(array): support max and min in array check (#22)
  * chore: use semantic-release workflow (#23)

2.1.0 / 2017-06-02
==================

  * feat: allowEmpty can be use with max and min
  * feat: allowEmpty can be use with format

2.0.0 / 2015-09-25
==================

 * test: use codecov.io
 * Refactor with Class, use ES6 class style.

1.2.0 / 2015-09-21
==================

 * feat: remove field name from error message. #12

1.1.1 / 2015-09-15
==================

 * fix: it should not translate the `field` key.

1.1.0 / 2015-09-15
==================

 * feat: I18n translate possible by override the `validate.translate` method.

1.0.2 / 2015-01-12
==================

 * feat: add url
 * fix: remove david badge

1.0.1 / 2015-01-09
==================

  * fix(readme): fix exmaple arguments

1.0.0 / 2015-01-09
==================

 * fix: ensure error message correct when add custom rules
 * feat: add password
 * feat: add email type
 * feat(addRule): support add custom rules
 * doc: fix readme and add example
 * fix(array): array itemType
 * feat(string): detault allowEmpty to false
 * fix:(test) fix test cases
 * refactor: rewrite

0.0.7 / 2013-08-09
==================

  * support Enum check fixed #1

0.0.6 / 2013-08-07
==================

  * Custom error message
  * remove 0.11 from travis

0.0.5 / 2013-08-06
==================

  * support RegExp rule

0.0.4 / 2013-07-01
==================

  * support isArray and isObject

0.0.3 / 2013-06-26
==================

  * fixed required=false bug
  * run again to vaild cache functions

0.0.2 / 2013-06-26
==================

  * improve performance

0.0.1 / 2013-06-26
==================

  * add test cases
  * first commit


example.js
var Parameter = require('./');

var rule = {
  name: 'string',
  age: {type: 'int', max: 200},
  gender: ['male', 'female'],
  working: 'boolean',
  salary: {type: 'number', min: 0},
  birthday: 'date',
  now: 'dateTime',
  id: 'id',
  childrens: {
    type: 'array',
    itemType: 'object',
    required: false,
    rule: {
      name: 'string',
      age: { type: 'int', customMsg: 'age must be an integer!!!!' },
      gender: ['male', 'female'],
      birthday: {type: 'date', required: false}
    }
  },
  mate: {
    type: 'object',
    required: false,
    rule: {
      name: 'string',
      age: 'int',
      gender: ['male', 'female'],
      birthday: {type: 'date', required: false}
    }
  }
};

var valid = {
  name: 'foo',
  gender: 'male',
  age: 30,
  working: true,
  salary: 10000.1,
  birthday: '1990-01-01',
  now: '2015-01-07 00:00:00',
  id: '052111',
  childrens: [{
    name: 'bar1',
    age: 1,
    gender: 'female',
    birthday: '2014-01-01'
  }, {
    name: 'bar2',
    age: 2,
    gender: 'male',
    birthday: '2013-01-01'
  }],
  mate: {
    name: 'hee',
    age: 29,
    gender: 'female'
  }
};

var invalid = {
  name: 1,
  gender: 'male1',
  age: 300,
  working: 'true',
  salary: '10000.1',
  birthday: '1990-01-011',
  id: '052111x',
  childrens: [{
    name: 'bar1',
    gender: 'female',
    birthday: '2014-01-01'
  }, {
    name: 'bar2',
    age: 2,
    birthday: '2013-01-01'
  }],
  mate: {
    age: 29,
    gender: 'female'
  }
};

var p = new Parameter();

console.log('valid object result: ', p.validate(rule, valid));
console.log('invalid object result: ', p.validate(rule, invalid));


lib/regexps.js
'use strict';

exports.DATE_TYPE_RE = /^\d{4}\-\d{2}\-\d{2}$/;

exports.DATETIME_TYPE_RE = /^\d{4}\-\d{2}\-\d{2} \d{2}:\d{2}:\d{2}$/;

exports.ID_RE = /^\d+$/;

// http://www.regular-expressions.info/email.html
exports.EMAIL_RE = /^[a-z0-9\!\#\$\%\&\'\*\+\/\=\?\^\_\`\{\|\}\~\-]+(?:\.[a-z0-9\!\#\$\%\&\'\*\+\/\=\?\^\_\`\{\|\}\~\-]+)*@(?:[a-z0-9](?:[a-z0-9\-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9\-]*[a-z0-9])?$/i;

exports.PASSWORD_RE = /^[\w\`\~\!\@\#\$\%\^\&\*\(\)\-\_\=\+\[\]\{\}\|\;\:\'\"\,\<\.\>\/\?]+$/;

// https://gist.github.com/dperini/729294
exports.URL_RE = /^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:\/\S*)?$/i;


index.js
'use strict';

var util = require('util');

/**
 * Regexps
 */

var DATE_TYPE_RE = /^\d{4}\-\d{2}\-\d{2}$/;
var DATETIME_TYPE_RE = /^\d{4}\-\d{2}\-\d{2} \d{2}:\d{2}:\d{2}$/;
var ID_RE = /^\d+$/;

// http://www.regular-expressions.info/email.html
var EMAIL_RE = /^[a-z0-9\!\#\$\%\&\'\*\+\/\=\?\^\_\`\{\|\}\~\-]+(?:\.[a-z0-9\!\#\$\%\&\'\*\+\/\=\?\^\_\`\{\|\}\~\-]+)*@(?:[a-z0-9](?:[a-z0-9\-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9\-]*[a-z0-9])?$/i;

var PASSWORD_RE = /^[\w\`\~\!\@\#\$\%\^\&\*\(\)\-\_\=\+\[\]\{\}\|\;\:\'\"\,\<\.\>\/\?]+$/;

// https://gist.github.com/dperini/729294
var URL_RE = /^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:\/\S*)?$/i;

/**
 * Parameter class
 * @class Parameter
 */
class Parameter {
  constructor(opts) {
    opts = opts || {};

    if (typeof opts.translate === 'function') {
      this.translate = opts.translate;
    }

    if (opts.validateRoot) this.validateRoot = true;
    if (opts.convert) this.convert = true;
    if (opts.widelyUndefined) this.widelyUndefined = true;
  }

  t() {
    var args = Array.prototype.slice.call(arguments);
    if (typeof this.translate === 'function') {
      return this.translate.apply(this, args);
    } else {
      return util.format.apply(util, args);
    }
  }

  /**
   * validate
   *
   * @param {Object} rules
   * @return {Object} obj
   * @api public
   */
  validate(rules, obj) {
    if (typeof rules !== 'object') {
      throw new TypeError('need object type rule');
    }

    if (this.validateRoot && (typeof obj !== 'object' || !obj)) {
      return [{
        message: this.t('the validated value should be a object'),
        code: this.t('invalid'),
        field: undefined,
      }];
    }

    var self = this;

    var errors = [];

    for (var key in rules) {
      var rule = formatRule(rules[key]);
      var value = obj[key];

      if (typeof value === 'string' && rule.trim === true) {
        value = obj[key] = value.trim();
      }

      // treat null / '' / NaN as undefined
      let widelyUndefined = this.widelyUndefined;
      if ('widelyUndefined' in rule) widelyUndefined = rule.widelyUndefined;
      if (widelyUndefined &&
        (value === '' || value === null || Number.isNaN(value))) {
        value = obj[key] = undefined;
      }

      var has = value !== null && value !== undefined;

      if (!has) {
        if (rule.required !== false) {
          errors.push({
            message: this.t('required'),
            field: key,
            code: this.t('missing_field')
          });
        }
        // support default value
        if ('default' in rule) {
          obj[key] = rule.default;
        }
        continue;
      }

      var checker = TYPE_MAP[rule.type];
      if (!checker) {
        throw new TypeError('rule type must be one of ' + Object.keys(TYPE_MAP).join(', ') +
          ', but the following type was passed: ' + rule.type);
      }

      convert(rule, obj, key, this.convert);
      var msg = checker.call(self, rule, obj[key], obj);
      if (typeof msg === 'string') {
        errors.push({
          message: msg,
          code: this.t('invalid'),
          field: key
        });
      }

      if (Array.isArray(msg)) {
        msg.forEach(function (e) {
          var dot = rule.type === 'object' ? '.' : '';
          e.field = key + dot + e.field;
          errors.push(e);
        });
      }
    }

    if (errors.length) {
      return errors;
    }
  }
};

/**
 * Module exports
 * @type {Function}
 */
module.exports = Parameter;

/**
 * add custom rule to global rules list.
 *
 * @param {String} type
 * @param {Function | RegExp} check
 * @param {Boolean} [override] - override exists rule or not, default is true
 * @param {String|Function} [convertType]
 * @api public
 */
Parameter.prototype.addRule = Parameter.addRule = function addRule(type, check, override, convertType) {
  if (!type) {
    throw new TypeError('`type` required');
  }

  // addRule(type, check, convertType)
  if (typeof override === 'string' || typeof override === 'function') {
    convertType = override;
    override = true;
  }

  if (typeof override !== 'boolean') {
    override = true;
  }

  if (!override && TYPE_MAP[type]) {
    throw new TypeError('rule `' + type + '` exists');
  }

  if (convertType) {
    if (typeof convertType !== 'string' && typeof convertType !== 'function') {
      throw new TypeError('convertType should be string or function');
    }
    Parameter.CONVERT_MAP[type] = convertType;
  }


  if (typeof check === 'function') {
    TYPE_MAP[type] = check;
    return;
  }

  if (check instanceof RegExp) {
    TYPE_MAP[type] = function (rule, value) {
      return checkString.call(this, {format: check}, value);
    };
    return;
  }

  throw new TypeError('check must be function or regexp');
};

/**
 * Simple type map
 * @type {Object}
 */
var TYPE_MAP = Parameter.TYPE_MAP = {
  number: checkNumber,
  int: checkInt,
  integer: checkInt,
  string: checkString,
  id: checkId,
  date: checkDate,
  dateTime: checkDateTime,
  datetime: checkDateTime,
  boolean: checkBoolean,
  bool: checkBoolean,
  array: checkArray,
  object: checkObject,
  enum: checkEnum,
  email: checkEmail,
  password: checkPassword,
  url: checkUrl,
};

var CONVERT_MAP = Parameter.CONVERT_MAP = {
  number: 'number',
  int: 'int',
  integer: 'int',
  string: 'string',
  id: 'string',
  date: 'string',
  dateTime: 'string',
  datetime: 'string',
  boolean: 'bool',
  bool: 'bool',
  email: 'string',
  password: 'string',
  url: 'string',
};

/**
 * format a rule
 * - resolve abbr
 * - resolve `?`
 * - resolve default convertType
 *
 * @param {Mixed} rule
 * @return {Object}
 * @api private
 */

function formatRule(rule) {
  rule = rule || {};
  if (typeof rule === 'string') {
    rule = { type: rule };
  } else if (Array.isArray(rule)) {
    rule = { type: 'enum', values: rule };
  } else if (rule instanceof RegExp) {
    rule = { type: 'string', format: rule };
  }

  if (rule.type && rule.type[rule.type.length - 1] === '?') {
    rule.type = rule.type.slice(0, -1);
    rule.required = false;
  }

  return rule;
}

/**
 * convert param to specific type
 * @param {Object} rule
 * @param {Object} obj
 * @param {String} key
 * @param {Boolean} defaultConvert
 */
function convert(rule, obj, key, defaultConvert) {
  var convertType;
  if (defaultConvert) convertType = CONVERT_MAP[rule.type];
  if (rule.convertType) convertType = rule.convertType;
  if (!convertType) return;

  const value = obj[key];
  // convert type only work for primitive data
  if (typeof value === 'object') return;

  // convertType support function
  if (typeof convertType === 'function') {
    obj[key] = convertType(value, obj);
    return;
  }

  switch (convertType) {
    case 'int':
      obj[key] = parseInt(value, 10);
      break;
    case 'string':
      obj[key] = String(value);
      break;
    case 'number':
      obj[key] = Number(obj[key]);
      break;
    case 'bool':
    case 'boolean':
      obj[key] = !!value;
      break;
    default:
      // support convertType function added by addRule
      if (typeof CONVERT_MAP[convertType] === 'function' ) {
        obj[key] = CONVERT_MAP[rule.type](obj[key]);
      }
  }
}

/**
 * check interger
 * {
 *   max: 100,
 *   min: 0
 * }
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkInt(rule, value) {
  if (typeof value !== 'number' || value % 1 !== 0) {
    return this.t('should be an integer');
  }

  if (rule.hasOwnProperty('max') && value > rule.max) {
    return this.t('should smaller than %s', rule.max);
  }

  if (rule.hasOwnProperty('min') && value < rule.min) {
    return this.t('should bigger than %s', rule.min);
  }
}

/**
 * check number
 * {
 *   max: 100,
 *   min: 0
 * }
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkNumber(rule, value) {
  if (typeof value !== 'number' || isNaN(value)) {
    return this.t('should be a number');
  }
  if (rule.hasOwnProperty('max') && value > rule.max) {
    return this.t('should smaller than %s', rule.max);
  }
  if (rule.hasOwnProperty('min') && value < rule.min) {
    return this.t('should bigger than %s', rule.min);
  }
}

/**
 * check string
 * {
 *   allowEmpty: true, // resolve default convertType to false, alias to empty)
 *   format: /^\d+$/,
 *   max: 100,
 *   min: 0,
 *   trim: false,
 * }
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkString(rule, value) {
  if (typeof value !== 'string') {
    return this.t('should be a string');
  }

  // if required === false, set allowEmpty to true by default
  if (!rule.hasOwnProperty('allowEmpty') && rule.required === false) {
    rule.allowEmpty = true;
  }

  var allowEmpty = rule.hasOwnProperty('allowEmpty')
    ? rule.allowEmpty
    : rule.empty;

  if (!value) {
    if (allowEmpty) return;
    return this.t('should not be empty');
  }

  if (rule.hasOwnProperty('max') && value.length > rule.max) {
    return this.t('length should smaller than %s', rule.max);
  }
  if (rule.hasOwnProperty('min') && value.length < rule.min) {
    return this.t('length should bigger than %s', rule.min);
  }

  if (rule.format && !rule.format.test(value)) {
    return rule.message || this.t('should match %s', rule.format);
  }
}

/**
 * check id format
 * format: /^\d+/
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkId(rule, value) {
  return checkString.call(this, { format: ID_RE, allowEmpty: rule.allowEmpty }, value);
}

/**
 * check date format
 * format: YYYY-MM-DD
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkDate(rule, value) {
  return checkString.call(this, { format: DATE_TYPE_RE, allowEmpty: rule.allowEmpty }, value);
}

/**
 * check date time format
 * format: YYYY-MM-DD HH:mm:ss
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkDateTime(rule, value) {
  return checkString.call(this, { format: DATETIME_TYPE_RE, allowEmpty: rule.allowEmpty }, value);
}

/**
 * check boolean
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkBoolean(rule, value) {
  if (typeof value !== 'boolean') {
    return this.t('should be a boolean');
  }
}

/**
 * check enum
 * {
 *   values: [0, 1, 2]
 * }
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkEnum(rule, value) {
  if (!Array.isArray(rule.values)) {
    throw new TypeError('check enum need array type values');
  }
  if (rule.values.indexOf(value) === -1) {
    return this.t('should be one of %s', rule.values.join(', '));
  }
}

/**
 * check email
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkEmail(rule, value) {
  return checkString.call(this, {
    format: EMAIL_RE,
    message: rule.message || this.t('should be an email'),
    allowEmpty: rule.allowEmpty,
  }, value);
}

/**
 * check password
 * @param {Object} rule
 * @param {Object} value
 * @param {Object} obj
 * @return {Boolean}
 *
 * @api private
 */

function checkPassword(rule, value, obj) {
  if (!rule.min) {
    rule.min = 6;
  }
  rule.format = PASSWORD_RE;
  var error = checkString.call(this, rule, value);
  if (error) {
    return error;
  }
  if (rule.compare && obj[rule.compare] !== value) {
    return this.t('should equal to %s', rule.compare);
  }
}

/**
 * check url
 *
 * @param {Object} rule
 * @param {Object} value
 * @return {Boolean}
 * @api private
 */

function checkUrl(rule, value) {
  return checkString.call(this, {
    format: URL_RE,
    message: rule.message || this.t('should be a url'),
    allowEmpty: rule.allowEmpty
  }, value);
}

/**
 * check object
 * {
 *   rule: {}
 * }
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkObject(rule, value) {
  if (typeof value !== 'object') {
    return this.t('should be an object');
  }

  if (rule.rule) {
    return this.validate(rule.rule, value);
  }
}

/**
 * check array
 * {
 *   type: 'array',
 *   itemType: 'string'
 *   rule: {type: 'string', allowEmpty: true}
 * }
 *
 * {
 *   type: 'array'.
 *   itemType: 'object',
 *   rule: {
 *     name: 'string'
 *   }
 * }
 *
 * @param {Object} rule
 * @param {Mixed} value
 * @return {Boolean}
 * @api private
 */

function checkArray(rule, value) {
  if (!Array.isArray(value)) {
    return this.t('should be an array');
  }

  if (rule.hasOwnProperty('max') && value.length > rule.max) {
    return this.t('length should smaller than %s', rule.max);
  }
  if (rule.hasOwnProperty('min') && value.length < rule.min) {
    return this.t('length should bigger than %s', rule.min);
  }

  if (!rule.itemType) {
    return;
  }

  var self = this;
  var checker = TYPE_MAP[rule.itemType];
  if (!checker) {
    throw new TypeError('rule type must be one of ' + Object.keys(TYPE_MAP).join(', ') +
        ', but the following type was passed: ' + rule.itemType);
  }

  var errors = [];
  var subRule = rule.itemType === 'object'
  ? rule
  : rule.rule || formatRule(rule.itemType);

  value.forEach(function (v, i) {
    var index = '[' + i + ']';
    var errs = checker.call(self, subRule, v);

    if (typeof errs === 'string') {
      errors.push({
        field: index,
        message: errs,
        code: self.t('invalid')
      });
    }
    if (Array.isArray(errs)) {
      errors = errors.concat(errs.map(function (e) {
        e.field = index + '.' + e.field;
        e.message = e.message;
        return e;
      }));
    }
  });

  return errors;
}
